### PR TITLE
Add summation in Q to ILL's reflectometry reduction

### LINKED ISSUE #22315: Coherent or SumInQ mode for ReflectometryILLSumForeground
ILL's reflectometry reduction workflow is missing the coherent summation method, a.k.a SumInQ mode. This should be added to the `ReflectometryILLSumForeground` algorithm. An implementation exists in `ReflectometryReductionOne` which could be extracted into a separate algorithm and shared between the reductions.

### PR DESCRIPTION
**Pull request by soininen**
This PR adds a previously missing 'SumInQ' option to `ReflectometryILLSumForeground`. The option enables summation in Q, also know as coherent summation.

There were a few modifications to the overal reduction workflow. Most importantly cropping in wavelength was moved to the end of `ReflectometryILLSumForeground` from `ReflectometryILLPreprocess`.

The wavelength ranges were removed from `ReflectometrySumInQ` and an option to exclude partial bins following #22541 was added.

The changes to algorithm properties and thus the API should not affect any users since none of the algorithms touched here have seen any use outside the development team and test user at ILL.

**Report to:** nobody.

**To test:**

The following script could be used as a starting point for testing. Note that the default facility should be set to ILL and the unit test data directories should be included in Mantid's data search directories before running the script.

```python
ReflectometryILLPreprocess(
    Run='ILL/D17/317369.nxs',
    OutputWorkspace='direct-raw',
    ForegroundHalfWidth=[3],
    OutputBeamPositionWorkspace='direct-beamPos')
ReflectometryILLPreprocess(
    Run='ILL/D17/317370.nxs',
    OutputWorkspace='reflected-raw',
    ForegroundHalfWidth=[3, 4],
    BeamPositionWorkspace='direct-beamPos')
ReflectometryILLSumForeground(
    InputWorkspace='direct-raw',
    OutputWorkspace='direct-fgd')
ReflectometryILLSumForeground(
    InputWorkspace='reflected-raw',
    OutputWorkspace='reflectivity',
    WavelengthRange=[2., 14.],
    SummationType='SumInQ',
    DirectForegroundWorkspace='direct-fgd')
```

Fixes #22315.

Does this update require release notes?
- [x] Yes
- [ ] No

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by vreimund — APPROVED** [review]
Tested reduction work-flow for reflectometry data and it works fine.

File: Framework/Algorithms/inc/MantidAlgorithms/ReflectometrySumInQ.h [hunk 1]
```diff
@@ -43,12 +43,20 @@ namespace Algorithms {
 */
 class MANTID_ALGORITHMS_DLL ReflectometrySumInQ : public API::Algorithm {
 public:
+  struct Angles {
+    double horizon{std::nan("")};
```
Comment (peterfpeterson) [inline Framework/Algorithms/inc/MantidAlgorithms/ReflectometrySumInQ.h]: Please change this to [`std::numeric_limits<double>::quiet_NaN()`](https://en.cppreference.com/w/cpp/types/numeric_limits/quiet_NaN).

File: Framework/Algorithms/src/ReflectometrySumInQ.cpp [hunk 2]
```diff
@@ -161,6 +169,28 @@ void ReflectometrySumInQ::MinMax::testAndSet(const double a) noexcept {
   }
 }
 
+/**
+* Set the `max` field if `a` is geater than `max`.
+*
+* @param a [in] :: a number
+*/
+void ReflectometrySumInQ::MinMax::testAndSetMax(const double a) noexcept {
+  if (a > max) {
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/ReflectometrySumInQ.cpp]: Isn't this just `max = std::max(a,max)`?

File: Framework/Algorithms/src/ReflectometrySumInQ.cpp [hunk 3]
```diff
@@ -161,6 +169,28 @@ void ReflectometrySumInQ::MinMax::testAndSet(const double a) noexcept {
...
+  }
+}
+
+/**
+* Set the `min` field if `a` is smaller than `min`.
+*
+* @param a [in] :: a number
+*/
+void ReflectometrySumInQ::MinMax::testAndSetMin(const double a) noexcept {
+  if (a < min) {
```
Comment (peterfpeterson) [inline Framework/Algorithms/src/ReflectometrySumInQ.cpp]: Isn't this just `min = std::min(a,min)`?

**Review by peterfpeterson — CHANGES_REQUESTED** [review]
A couple of super-small changes and I'll merge it.

**Review by peterfpeterson — APPROVED** [review]

**Pull request merged**